mesonpy/__init__.py

```diff
@@ -566,13 +566,41 @@ def __init__(
         self._build_dir = pathlib.Path(build_dir).absolute() if build_dir else (self._working_dir / 'build')
         self._install_dir = self._working_dir / 'install'
         self._meson_native_file = self._source_dir / '.mesonpy-native-file.ini'
+        self._meson_cross_file = self._source_dir / '.mesonpy-cross-file.ini'
+        self._meson_args: MesonArgs = collections.defaultdict(list)
         self._env = os.environ.copy()
 
         # prepare environment
         ninja_path = _env_ninja_command()
         if ninja_path is not None:
             self._env.setdefault('NINJA', str(ninja_path))
 
+        # setuptools-like ARCHFLAGS environment variable support
+        if sysconfig.get_platform().startswith('macosx-'):
+            archflags = self._env.get('ARCHFLAGS')
+            if archflags is not None:
+                arch, *other = filter(None, (x.strip() for x in archflags.split('-arch')))
+                if other:
+                    raise ConfigError(f'multi-architecture builds are not supported but $ARCHFLAGS={archflags!r}')
+                macver, _, nativearch = platform.mac_ver()
+                if arch != nativearch:
+                    x = self._env.setdefault('_PYTHON_HOST_PLATFORM', f'macosx-{macver}-{arch}')
+                    if not x.endswith(arch):
+                        raise ConfigError(f'$ARCHFLAGS={archflags!r} and $_PYTHON_HOST_PLATFORM={x!r} do not agree')
+                    family = 'aarch64' if arch == 'arm64' else arch
+                    cross_file_data = textwrap.dedent(f'''
+                        [binaries]
+                        c = ['cc', '-arch', {arch!r}]
+                        cpp = ['cpp', '-arch', {arch!r}]
+                        [host_machine]
+                        system = 'Darwin'
+                        cpu = {arch!r}
+                        cpu_family = {family!r}
+                        endian = 'little'
+                    ''')
+                    self._meson_cross_file.write_text(cross_file_data)
+                    self._meson_args['setup'].extend(('--cross-file', os.fspath(self._meson_cross_file)))
```

Comment:
Humm, I'd prefer the cross file not to be added to `meson_args`, but rather added in `_configure` directly. I think it would be alright to always have a cross file, and it just being empty when it is not needed. I feel like that is cleaner and better sets things up for the future if we need to use it for something else.

Reply:
Adding `--cross-file` for a non cross build sounds more confusing than helpful to me. I'd much prefer to not see that.

Reply:
I agree with @rgommers Also, at the moment I cannot imagine any other feature that would automagically generate a cross-file. If we will find one, we will revisit this. One main advantage of not having the `--cross-file` argument when not needed, is that its presence when scanning the build output is already a good indication that the environment variables are set correctly.

Reply:
This could still be done in optionally `_configure`, without hijacking the user args.